File: services/agent/build-modules/line04/wechat-rpa/panel/panel_events.py
"""
panel_events.py — 作战窗 Agent Panel 刀1 独立事件写入
PrepPRD: sprints/07280929-agent-panel-knife1/prep-prd.md

与 line04 现有 events.jsonl（唯一写者=listen_chat.py，供画像卡 overlay_window.py 只读消费）
完全独立的新文件 panel-events.jsonl，绝不能共用——否则破坏现有单写者约束
（golden-path-4-smoke.sh Step15c 曾因单写者被破坏出过真机事故：画像卡永远显示不出内容）。

6 种事件：task_started / step / waiting / stuck / done / failed
"""
import json
import os
import time
import uuid

from overlay.pii_filter import filter_pii

_VALID_EVENTS = {"task_started", "step", "waiting", "stuck", "done", "failed"}

_PANEL_EVENTS_FILENAME = "panel-events.jsonl"


def _state_dir():
    return os.environ.get("ZJ_STATE_DIR") or os.environ.get("PUBLIC", r"C:\Users\Public")


def _panel_events_path():
    return os.path.join(_state_dir(), _PANEL_EVENTS_FILENAME)


def new_task_id(sender):
    return f"{sender}-{int(time.time() * 1000)}-{uuid.uuid4().hex[:6]}"


def write_panel_event(event, task_id, line, device, title, detail=None, progress=None, severity="info"):
    """追加写入一条作战窗事件（软失败：任何异常仅静默吞，不上抛——面板是旁观者，
    绝不能因为写事件失败反过来影响 handler 干活，参照 shared/event-reporter.ts 纪律）。"""
    if event not in _VALID_EVENTS:
        raise ValueError(f"unknown panel event type: {event!r}, must be one of {sorted(_VALID_EVENTS)}")

    record = {
        "event": event,
        "task_id": task_id,
        "line": line,
        "device": device,
        "title": filter_pii(title) if title else title,
        "detail": filter_pii(detail) if detail else detail,
        "progress": list(progress) if progress else None,
        "severity": severity,
        "ts": time.time(),
    }
    try:
        with open(_panel_events_path(), "a", encoding="utf-8") as f:
            f.write(json.dumps(record, ensure_ascii=False) + "\n")
    except OSError:
        pass

File: apps/api/src/routes/panel-events.ts
/**
 * /api/panel/events — 作战窗 Agent Panel 刀1 薄写入端点
 *
 * PrepPRD: sprints/07280929-agent-panel-knife1/prep-prd.md
 * 鉴权：X-Internal-Token（内部Agent调用，env未设置时dev放行）+ X-Tenant-Id（租户隔离，必填）。
 * 中台侧写入失败的重试/本地缓存策略在 Agent 侧（判定点已登记），本端点本身只做简单 INSERT。
 */

import { Router, Request } from 'express';
import { internalAuth } from '../middleware/internal-auth';
import { writePanelEvent, PanelEventType } from '../services/panel/panel-events-service';

export const panelEventsRouter = Router();

panelEventsRouter.use(internalAuth);

const VALID_EVENTS: PanelEventType[] = ['task_started', 'step', 'waiting', 'stuck', 'done', 'failed'];

const MISSING_TENANT = {
  error: 'MISSING_TENANT',
  message: '缺 tenant_id（X-Tenant-Id 头）— 拒绝执行，不回退全量',
};

function resolveTenantId(req: Request): string {
  return (req.header('X-Tenant-Id') || '').trim();
}

panelEventsRouter.post('/events', async (req, res) => {
  const tenantId = resolveTenantId(req);
  if (!tenantId) return res.status(400).json(MISSING_TENANT);

  const {
    event, task_id: taskId, line, device, title, detail, progress, severity,
  } = req.body || {};

  if (!event || !taskId || !line || !device || !title) {
    return res.status(400).json({
      error: 'MISSING_FIELDS',
      message: '缺 event/task_id/line/device/title',
    });
  }
  if (!VALID_EVENTS.includes(event)) {
    return res.status(400).json({
      error: 'INVALID_EVENT',
      message: `event 必须是 ${VALID_EVENTS.join('/')} 之一`,
    });
  }

  try {
    const r = await writePanelEvent({
      tenantId,
      event,
      taskId,
      line,
      device,
      title,
      detail: detail ?? null,
      progress: progress ?? null,
      severity: severity ?? 'info',
    });
    return res.status(200).json({ ok: true, id: r.id });
  } catch (e) {
    console.error('[panel/events] error:', e);
    return res.status(500).json({ error: 'INTERNAL', message: 'write failed' });
  }
});
